drgn/helpers/linux/slab.py

```diff
@@ -53,3 +55,163 @@ def print_slab_caches(prog: Program) -> None:
+    for p in for_each_page(prog):
+        try:
+            if p.flags.value_() & PGSlab:
+                yield p
+        except FaultError:
+            pass
+
+
+def _get_slab_objects_slab_as_type(
+    prog: Program,
+    slab_cache: Object,
+    obj_type: Type,
+    slab_type: Type,
+    page_objects_func: Callable[[Program, Object, Object, Type], List[Object]],
+) -> List[Object]:
+    ret = []
+    for page in for_each_slab_page(prog):
+        slab = cast(slab_type, page)
+        if slab.slab_cache == slab_cache:
+            ret.extend(page_objects_func(prog, slab_cache, slab, obj_type))
+    return ret
+
+
+def _get_slab_objects_legacy(
+    prog: Program,
+    slab_cache: Object,
+    obj_type: Type,
+    page_objects_func: Callable[[Program, Object, Object, Type], List[Object]],
+) -> List[Object]:
+    ret = []
+    for page in for_each_slab_page(prog):
+        if page.slab_cache == slab_cache:
+            ret.extend(page_objects_func(prog, slab_cache, page, obj_type))
+    return ret
+
+
+def get_allocated_slab_objects(
+    prog: Program, slab_cache: Object, obj_type: Type
+) -> List[Object]:
+    """
+    Get the list of all allocated objects of a given slab and type
+
+    >>> dentry_slab = find_slab_cache(prog, "dentry")
+    >>> dentry_type = prog.type("struct dentry")
+    >>> get_allocated_slab_objects(prog, dentry_slab, dentry_type)
+    [Object(prog, 'struct dentry', address=0xffff888c8417de00), Object(prog, 'struct dentry', address=0xffff888c8417dec0)]
+
+    :param slab_cache: ``struct kmem_cache``
+    :param obj_type: type of object contained in the slab
+    """
+    # Linux kernel commit d122019bf061cccc4583eb9ad40bf58c2fe517be ("mm: Split slab into its own type")
+    # (in v5.17) moved slab fields out of struct page into struct slab and allows page to be cast to slab
+    slub = False
+    try:
+        freelist_index_size = sizeof(prog.type("freelist_idx_t"))
+    except LookupError:
+        slub = True
+
+    red_left_pad = 0
+    if prog.type("struct kmem_cache").has_member("red_left_pad"):
+        red_left_pad += slab_cache.red_left_pad.value_()
+
+    def _slub_page_objects(
+        prog: Program, slab_cache: Object, page_or_slab: Object, obj_type: Type
+    ) -> List[Object]:
+        """
+        Get the list of all objects from a slub page
+        :param slab_cache: ``struct kmem_cache``
+        :param page_or_slab: ``struct page`` or ``struct slab`` depending on the kernel version
+        :param obj_type: type of object contained in the slab_cache
+        """
+        addr = page_to_virt(cast(prog.type("struct page *"), page_or_slab)).value_()
+        if not isinstance(addr, int):
+            addr = addr.value_()
+        addr += red_left_pad
+
+        end = addr + slab_cache.size * page_or_slab.objects
+        ret = []
+
+        slab_cache_size = slab_cache.size.value_()
+        freelist_set = obtain_freelist_slub(page_or_slab, slab_cache.offset.value_())
```

Comment:
Get `slab_cache.offset.value_()` once outside of the function (maybe name it `freelist_offset`? The kernel naming is very ambiguous.)